contracts/misc/P2PPairStaking.sol

```diff
@@ -0,0 +1,586 @@
+
+        emit OrderCancelled(orderHash, listingOrder.offerer);
+    }
+
+    function matchPairStakingList(
+        ListingOrder calldata apeOrder,
+        ListingOrder calldata apeCoinOrder
+    ) external returns (bytes32 orderHash) {
+        //1 validate all order
+        _validateApeOrder(apeOrder);
+        _validateApeCoinOrder(apeCoinOrder);
+
+        //2 check if orders can match
+        require(
+            apeOrder.stakingType == StakingType.BAYCStaking ||
+                apeOrder.stakingType == StakingType.MAYCStaking,
+            "invalid stake type"
+        );
+        require(
+            apeOrder.stakingType == apeCoinOrder.stakingType,
+            "orders type match failed"
+        );
+        require(
+            apeOrder.share + apeCoinOrder.share ==
+                PercentageMath.PERCENTAGE_FACTOR,
+            "orders share match failed"
+        );
+
+        //3 transfer token
+        IERC721(apeOrder.token).safeTransferFrom(
+            apeOrder.offerer,
+            address(this),
+            apeOrder.tokenId
+        );
+        uint256 apeAmount = getApeCoinStakingCap(apeOrder.stakingType);
+        IERC20(apeCoin).safeTransferFrom(
+            apeCoinOrder.offerer,
+            address(this),
+            apeAmount
+        );
+
+        //4 create match order
+        MatchedOrder memory matchedOrder;
+        matchedOrder.stakingType = apeOrder.stakingType;
+        matchedOrder.apeToken = apeOrder.token;
+        matchedOrder.apeOfferer = apeOrder.offerer;
+        matchedOrder.apeTokenId = apeOrder.tokenId;
+        matchedOrder.apeShare = apeOrder.share;
+        matchedOrder.apeCoinOfferer = apeCoinOrder.offerer;
+        matchedOrder.apeCoinShare = apeCoinOrder.share;
+        matchedOrder.apePrincipleAmount = apeAmount;
+        orderHash = getMatchedOrderHash(matchedOrder);
+        matchedOrders[orderHash] = matchedOrder;
+
+        //5 stake for ApeCoinStaking
+        ApeCoinStaking.SingleNft[]
+            memory singleNft = new ApeCoinStaking.SingleNft[](1);
+        singleNft[0].tokenId = apeOrder.tokenId;
+        singleNft[0].amount = apeAmount.toUint224();
+        if (apeOrder.stakingType == StakingType.BAYCStaking) {
+            apeCoinStaking.depositBAYC(singleNft);
+        } else {
+            apeCoinStaking.depositMAYC(singleNft);
+        }
+
+        //5 emit event
+        emit PairStakingMatched(orderHash);
+
+        return orderHash;
+    }
+
+    function matchBAKCPairStakingList(
+        ListingOrder calldata apeOrder,
+        ListingOrder calldata bakcOrder,
+        ListingOrder calldata apeCoinOrder
+    ) external returns (bytes32 orderHash) {
+        //1 validate all order
+        _validateApeOrder(apeOrder);
+        _validateBakcOrder(bakcOrder);
+        _validateApeCoinOrder(apeCoinOrder);
+
+        //2 check if orders can match
+        require(
+            apeOrder.stakingType == StakingType.BAYCPairStaking ||
```

Comment:
shall we be consistent on using `apeOrder.stakingType >= StakingType.BAYCPairStaking`

Reply:
ok, it's more simpler. I'll change it.